File: Src/java/cql-to-elm/src/jvmTest/java/org/cqframework/cql/cql2elm/operators/StringOperatorsTest.java
package org.cqframework.cql.cql2elm.operators;

import static org.cqframework.cql.cql2elm.matchers.HasTypeAndResult.hasTypeAndResult;
import static org.cqframework.cql.cql2elm.matchers.ListOfLiterals.listOfLiterals;
import static org.cqframework.cql.cql2elm.matchers.LiteralFor.literalFor;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import org.cqframework.cql.cql2elm.*;
import org.cqframework.cql.cql2elm.CqlCompilerException.ErrorSeverity;
import org.cqframework.cql.cql2elm.LibraryBuilder.SignatureLevel;
import org.hl7.elm.r1.*;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

class StringOperatorsTest {

    private static Map<String, ExpressionDef> defs;

    @BeforeAll
    static void setup() throws IOException {
        ModelManager modelManager = new ModelManager();
        CqlTranslator translator = CqlTranslator.fromStream(
                StringOperatorsTest.class.getResourceAsStream("../OperatorTests/StringOperators.cql"),
                new LibraryManager(modelManager, new CqlCompilerOptions(ErrorSeverity.Warning, SignatureLevel.None)));
        assertThat(translator.getErrors().size(), is(0));
        Library library = translator.toELM();
        defs = new HashMap<>();
        for (ExpressionDef def : library.getStatements().getDef()) {
            defs.put(def.getName(), def);
        }
    }

    @Test
    void add() {
        ExpressionDef def = defs.get("StringAdd");
        assertThat(def, hasTypeAndResult(Concatenate.class, "System.String"));
    }

    @Test
    void concatenate() {
        ExpressionDef def = defs.get("StringConcatenate");
        assertThat(def, hasTypeAndResult(Concatenate.class, "System.String"));
    }

    @Test
    void concatenateWithAmpersand() {
        ExpressionDef def = defs.get("StringConcatenateWithAmpersand");
        assertThat(def, hasTypeAndResult(Concatenate.class, "System.String"));

        Concatenate concatenate = (Concatenate) def.getExpression();
        for (Expression operand : concatenate.getOperand()) {
            assertThat(operand.getClass() == Coalesce.class, is(true));
        }
    }

    @Test
    void combine() {
        ExpressionDef def = defs.get("StringCombine");
        assertThat(def, hasTypeAndResult(Combine.class, "System.String"));
        Combine combine = (Combine) def.getExpression();
        assertThat(combine.getSource(), listOfLiterals("First", "Second", "Third", "Fourth"));
        assertThat(combine.getSeparator(), literalFor(","));

        def = defs.get("StringCombineNoSeparator");
        assertThat(def, hasTypeAndResult(Combine.class, "System.String"));
        combine = (Combine) def.getExpression();
        assertThat(combine.getSource(), listOfLiterals("abc", "def", "ghi", "jkl"));
        assertThat(combine.getSeparator(), nullValue());
    }

    @Test
    void split() {
        ExpressionDef def = defs.get("StringSplit");
        assertThat(def, hasTypeAndResult(Split.class, "list<System.String>"));
        Split split = (Split) def.getExpression();
        assertThat(split.getStringToSplit(), literalFor("First,Second,Third,Fourth"));
        assertThat(split.getSeparator(), literalFor(","));
    }

    @Test
    void splitOnMatches() {
        ExpressionDef def = defs.get("StringSplitOnMatches");
        assertThat(def, hasTypeAndResult(SplitOnMatches.class, "list<System.String>"));
        SplitOnMatches splitOnMatches = (SplitOnMatches) def.getExpression();
        assertThat(splitOnMatches.getStringToSplit(), literalFor("First,Second,Third,Fourth"));
        assertThat(splitOnMatches.getSeparatorPattern(), literalFor(","));
    }

    @Test
    void upper() {
        ExpressionDef def = defs.get("StringUpper");
        assertThat(def, hasTypeAndResult(Upper.class, "System.String"));
        Upper upper = (Upper) def.getExpression();
        assertThat(upper.getOperand(), literalFor("John"));
    }

    @Test
    void lower() {
        ExpressionDef def = defs.get("StringLower");
        assertThat(def, hasTypeAndResult(Lower.class, "System.String"));
        Lower lower = (Lower) def.getExpression();
        assertThat(lower.getOperand(), literalFor("John"));
    }

    @Test
    void positionOf() {
        ExpressionDef def = defs.get("StringPositionOf");
        assertThat(def, hasTypeAndResult(PositionOf.class, "System.Integer"));
        PositionOf positionOf = (PositionOf) def.getExpression();
        assertThat(positionOf.getPattern(), literalFor("J"));
        assertThat(positionOf.getString(), literalFor("John"));
    }

    @Test
    void lastPositionOf() {
        ExpressionDef def = defs.get("StringLastPositionOf");
        assertThat(def, hasTypeAndResult(LastPositionOf.class, "System.Integer"));
        LastPositionOf lastPositionOf = (LastPositionOf) def.getExpression();
        assertThat(lastPositionOf.getPattern(), literalFor("J"));
        assertThat(lastPositionOf.getString(), literalFor("John"));
    }

    @Test
    void substring() {
        ExpressionDef def = defs.get("StringSubstring");
        assertThat(def, hasTypeAndResult(Substring.class, "System.String"));
        Substring substring = (Substring) def.getExpression();
        // Note: these casts to Expression are necessary because of bug in expression.xsd (DSTU comment #824)
        assertThat((Expression) substring.getStringToSub(), literalFor("JohnDoe"));
        assertThat((Expression) substring.getStartIndex(), literalFor(5));

        def = defs.get("StringSubstringWithLength");
        assertThat(def, hasTypeAndResult(Substring.class, "System.String"));
        substring = (Substring) def.getExpression();
        assertThat((Expression) substring.getStringToSub(), literalFor("JohnDoe"));
        assertThat((Expression) substring.getStartIndex(), literalFor(1));
        assertThat((Expression) substring.getLength(), literalFor(4));
    }

    @Test
    void length() {
        ExpressionDef def = defs.get("StringLength");
        assertThat(def, hasTypeAndResult(Length.class, "System.Integer"));

        Length length = (Length) def.getExpression();
        assertThat(length.getOperand(), literalFor("John"));
    }

    @Test
    void startsWith() {
        ExpressionDef def = defs.get("StringStartsWith");
        assertThat(def, hasTypeAndResult(StartsWith.class, "System.Boolean"));
    }

    @Test
    void endsWith() {
        ExpressionDef def = defs.get("StringEndsWith");
        assertThat(def, hasTypeAndResult(EndsWith.class, "System.Boolean"));
    }

    @Test
    void matches() {
        ExpressionDef def = defs.get("StringMatches");
        assertThat(def, hasTypeAndResult(Matches.class, "System.Boolean"));
    }

    @Test
    void replaceMatches() {
        ExpressionDef def = defs.get("StringReplaceMatches");
        assertThat(def, hasTypeAndResult(ReplaceMatches.class, "System.String"));
    }
}

File: cql-to-elm/src/jvmTest/kotlin/org/cqframework/cql/cql2elm/operators/StringOperatorsTest.kt
package org.cqframework.cql.cql2elm.operators

import java.io.IOException
import kotlinx.io.asSource
import kotlinx.io.buffered
import org.cqframework.cql.cql2elm.CqlCompilerException
import org.cqframework.cql.cql2elm.CqlCompilerOptions
import org.cqframework.cql.cql2elm.CqlTranslator.Companion.fromSource
import org.cqframework.cql.cql2elm.LibraryBuilder.SignatureLevel
import org.cqframework.cql.cql2elm.LibraryManager
import org.cqframework.cql.cql2elm.ModelManager
import org.cqframework.cql.cql2elm.matchers.HasTypeAndResult.Companion.hasTypeAndResult
import org.cqframework.cql.cql2elm.matchers.ListOfLiterals.Companion.listOfLiterals
import org.cqframework.cql.cql2elm.matchers.LiteralFor.Companion.literalFor
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers
import org.hamcrest.Matchers.`is`
import org.hl7.elm.r1.Coalesce
import org.hl7.elm.r1.Combine
import org.hl7.elm.r1.Concatenate
import org.hl7.elm.r1.EndsWith
import org.hl7.elm.r1.Expression
import org.hl7.elm.r1.ExpressionDef
import org.hl7.elm.r1.LastPositionOf
import org.hl7.elm.r1.Length
import org.hl7.elm.r1.Lower
import org.hl7.elm.r1.Matches
import org.hl7.elm.r1.PositionOf
import org.hl7.elm.r1.ReplaceMatches
import org.hl7.elm.r1.Split
import org.hl7.elm.r1.SplitOnMatches
import org.hl7.elm.r1.StartsWith
import org.hl7.elm.r1.Substring
import org.hl7.elm.r1.Upper
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

internal class StringOperatorsTest {
    @Test
    fun add() {
        val def: ExpressionDef? = defs!!["StringAdd"]
        assertThat<ExpressionDef?>(def, hasTypeAndResult(Concatenate::class.java, "System.String"))
    }

    @Test
    fun concatenate() {
        val def: ExpressionDef? = defs!!["StringConcatenate"]
        assertThat<ExpressionDef?>(def, hasTypeAndResult(Concatenate::class.java, "System.String"))
    }

    @Test
    fun concatenateWithAmpersand() {
        val def: ExpressionDef = defs!!["StringConcatenateWithAmpersand"]!!
        assertThat(def, hasTypeAndResult(Concatenate::class.java, "System.String"))

        val concatenate = def.expression as Concatenate?
        for (operand in concatenate!!.operand) {
            assertThat(operand.javaClass === Coalesce::class.java, `is`(true))
        }
    }

    @Test
    fun combine() {
        var def: ExpressionDef = defs!!["StringCombine"]!!
        assertThat(def, hasTypeAndResult(Combine::class.java, "System.String"))
        var combine = def.expression as Combine?
        assertThat<Expression?>(
            combine!!.source,
            listOfLiterals("First", "Second", "Third", "Fourth"),
        )
        assertThat<Expression?>(combine.separator, literalFor(","))

        def = defs!!["StringCombineNoSeparator"]!!
        assertThat(def, hasTypeAndResult(Combine::class.java, "System.String"))
        combine = def.expression as Combine?
        assertThat<Expression?>(combine!!.source, listOfLiterals("abc", "def", "ghi", "jkl"))
        assertThat<Expression?>(combine.separator, Matchers.nullValue())
    }

    @Test
    fun split() {
        val def: ExpressionDef = defs!!["StringSplit"]!!
        assertThat(def, hasTypeAndResult(Split::class.java, "list<System.String>"))
        val split = def.expression as Split?
        assertThat<Expression?>(split!!.stringToSplit, literalFor("First,Second,Third,Fourth"))
        assertThat<Expression?>(split.separator, literalFor(","))
    }

    @Test
    fun splitOnMatches() {
        val def: ExpressionDef = defs!!["StringSplitOnMatches"]!!
        assertThat(def, hasTypeAndResult(SplitOnMatches::class.java, "list<System.String>"))
        val splitOnMatches = def.expression as SplitOnMatches?
        assertThat<Expression?>(
            splitOnMatches!!.stringToSplit,
            literalFor("First,Second,Third,Fourth"),
        )
        assertThat<Expression?>(splitOnMatches.separatorPattern, literalFor(","))
    }

    @Test
    fun upper() {
        val def: ExpressionDef = defs!!["StringUpper"]!!
        assertThat(def, hasTypeAndResult(Upper::class.java, "System.String"))
        val upper = def.expression as Upper?
        assertThat<Expression?>(upper!!.operand, literalFor("John"))
    }

    @Test
    fun lower() {
        val def: ExpressionDef = defs!!["StringLower"]!!
        assertThat(def, hasTypeAndResult(Lower::class.java, "System.String"))
        val lower = def.expression as Lower?
        assertThat<Expression?>(lower!!.operand, literalFor("John"))
    }

    @Test
    fun positionOf() {
        val def: ExpressionDef = defs!!["StringPositionOf"]!!
        assertThat(def, hasTypeAndResult(PositionOf::class.java, "System.Integer"))
        val positionOf = def.expression as PositionOf?
        assertThat<Expression?>(positionOf!!.pattern, literalFor("J"))
        assertThat<Expression?>(positionOf.string, literalFor("John"))
    }

    @Test
    fun lastPositionOf() {
        val def: ExpressionDef = defs!!["StringLastPositionOf"]!!
        assertThat(def, hasTypeAndResult(LastPositionOf::class.java, "System.Integer"))
        val lastPositionOf = def.expression as LastPositionOf?
        assertThat<Expression?>(lastPositionOf!!.pattern, literalFor("J"))
        assertThat<Expression?>(lastPositionOf.string, literalFor("John"))
    }

    @Test
    fun substring() {
        var def: ExpressionDef = defs!!["StringSubstring"]!!
        assertThat(def, hasTypeAndResult(Substring::class.java, "System.String"))
        var substring = def.expression as Substring?
        // Note: these casts to Expression are necessary because of bug in expression.xsd (DSTU
        // comment #824)
        assertThat<Expression?>(substring!!.stringToSub, literalFor("JohnDoe"))
        assertThat<Expression?>(substring.startIndex, literalFor(5))

        def = defs!!["StringSubstringWithLength"]!!
        assertThat(def, hasTypeAndResult(Substring::class.java, "System.String"))
        substring = def.expression as Substring?
        assertThat<Expression?>(substring!!.stringToSub, literalFor("JohnDoe"))
        assertThat<Expression?>(substring.startIndex, literalFor(1))
        assertThat<Expression?>(substring.length, literalFor(4))
    }

    @Test
    fun length() {
        val def: ExpressionDef = defs!!["StringLength"]!!
        assertThat(def, hasTypeAndResult(Length::class.java, "System.Integer"))

        val length = def.expression as Length?
        assertThat<Expression?>(length!!.operand, literalFor("John"))
    }

    @Test
    fun startsWith() {
        val def: ExpressionDef? = defs!!["StringStartsWith"]
        assertThat<ExpressionDef?>(def, hasTypeAndResult(StartsWith::class.java, "System.Boolean"))
    }

    @Test
    fun endsWith() {
        val def: ExpressionDef? = defs!!["StringEndsWith"]
        assertThat<ExpressionDef?>(def, hasTypeAndResult(EndsWith::class.java, "System.Boolean"))
    }

    @Test
    fun matches() {
        val def: ExpressionDef? = defs!!["StringMatches"]
        assertThat<ExpressionDef?>(def, hasTypeAndResult(Matches::class.java, "System.Boolean"))
    }

    @Test
    fun replaceMatches() {
        val def: ExpressionDef? = defs!!["StringReplaceMatches"]
        assertThat<ExpressionDef?>(
            def,
            hasTypeAndResult(ReplaceMatches::class.java, "System.String"),
        )
    }

    companion object {
        private var defs: MutableMap<String?, ExpressionDef>? = null

        @JvmStatic
        @BeforeAll
        @Throws(IOException::class)
        fun setup() {
            val modelManager = ModelManager()
            val translator =
                fromSource(
                    StringOperatorsTest::class
                        .java
                        .getResourceAsStream("../OperatorTests/StringOperators.cql")!!
                        .asSource()
                        .buffered(),
                    LibraryManager(
                        modelManager,
                        CqlCompilerOptions(
                            CqlCompilerException.ErrorSeverity.Warning,
                            SignatureLevel.None,
                        ),
                    ),
                )
            assertThat(translator.errors.size, `is`(0))
            val library = translator.toELM()
            defs = HashMap()
            for (def in library!!.statements!!.def) {
                defs!![def.name] = def
            }
        }
    }
}
